students/johnpharmd/lesson04/trigrams.py

```diff
@@ -28,30 +28,86 @@
 pron_set = set(['I', 'you', 'she', 'he', 'it', 'we', 'they'])
-
+n = 0
+new_st_lst= []
+testfile_dict = {1: 'sherlock_small.txt', 2: 'sherlock.txt'}
 
 # Processing
-st_split_lst = sample_st.split(' ')
-print('st_split_lst: ', st_split_lst, '\n')
-for i, word in enumerate(st_split_lst):
-    if i + 2 < len(st_split_lst):
-        s1 = word
-        s2 = st_split_lst[i + 1]
-        s3 = st_split_lst[i + 2]
-        st_set.update([s1, s2])
-        st_dict[count] = {frozenset(st_set): s3}
-        # print('st_set:', st_set)
-        count += 1
-        s1 = ''
-        s2 = ''
-        s3 = ''
-        st_set = set()
-for k, v in st_dict.items():
-    print(k, v, '\n')
+def clean_data(choice=1):
+    # use 2 as param, if want to test with 'sherlock.txt'
+    with open(testfile_dict[choice], 'r') as f:
+    # header_size = 0
+    # with open('sherlock.txt') as f:
+        read_data = f.read()
+        # read_data = f.seek('To Sherlock Holmes she')
+    read_data = ''.join(read_data.split('\r'))
+    read_data = ''.join(read_data.split('"'))
+    read_data = ' '.join(read_data.split('\n'))
+    read_data = ' '.join(read_data.split('--'))
+    read_data = ' '.join(read_data.split('-'))
+    read_data = ' '.join(read_data.split(', '))
+    read_data = ' '.join(read_data.split('. '))
+    read_data = ''.join(read_data.split('('))
+    read_data = ''.join(read_data.split(')'))
+    read_data = read_data.strip('.')
+    read_data = read_data.strip('  ')
+    read_data = read_data.split(' ')
+
+    print('type(read_data):', type(read_data), '\n')
+    # print('read_data:', read_data)
+    print(read_data)
+    return read_data
+
+
+def make_st_dict(lst):
+# lines between this and next comment line can be encapsulated into a
+# separate fxn
+# possible name: def make_st_dict(lst):
+    for i, word in enumerate(read_data):
+        if i + 2 < len(read_data):
+            s1 = word
+            s2 = read_data[i + 1]
+            st_key = (s1, s2)
+            s3 = read_data[i + 2]
+            if st_key not in st_dict:
+                st_dict[st_key] = [s3]
+            else:
+                st_dict[st_key] += [s3]
+        # count += 1
+    print(st_dict)
+# see comment line above
+
+
+def print_st_dict(dictionary):
+    print('\nst_dict:')
+    for k, v in st_dict.items():
+        print(k, v, end='\n')
+    print('len(st_dict):', len(st_dict))
+
+
+def make_new_st():
+    rand_k = random.choice(list(st_dict))
+    rand_v = random.choice(st_dict[rand_k])
+
+    new_st_lst.extend([rand_k[0], rand_k[1], rand_v])
+    print('initial new_st_lst:', new_st_lst)
+    potential_k = (new_st_lst[-2], new_st_lst[-1])
+    while potential_k in st_dict:  # len(st_dict) * 2:
+        # if potential_k in st_dict:
+        new_st_lst.append(random.choice(st_dict[potential_k]))
+        potential_k = (new_st_lst[-2], new_st_lst[-1])
+        n += 1    
+    print('len(new_st_lst):', len(new_st_lst))
+    print(new_st_lst)
+
+
+clean_data()
+make_st_dict(read_data)
```

Comment:
prefer this pattern:
```
cleaned_data = clean_data()
make_st_dict(cleaned_data)
```

especially since you are returning it in `clean_data` already